File: 11/SymbolTable.java
package comsys.compiler;

import comsys.compiler.enums.Range;
import comsys.compiler.identifier.Identifier;
import comsys.compiler.enums.Kind;

import java.util.HashMap;

public class SymbolTable {

    // 키 : 식별자 이름
    public static HashMap<String, Identifier> classRangeTable = new HashMap<>();
    public static HashMap<String, Identifier> subroutineRangeTable = new HashMap<>();

    public SymbolTable() {}

    public static void init() {
        classRangeTable.clear();
        subroutineRangeTable.clear();
        Kind.initAllIndex();
    }

    // 새로운 서브루틴 범위 시작
    public static void startSubroutine() {
        subroutineRangeTable.clear();
        Kind.ARG.setIndex(0);
        Kind.VAR.setIndex(0);
    }

    public static void define(String name, String type, Kind kind) {
        Identifier identifier = new Identifier(name, type, kind, kind.getIndex());
        kind.addIndex(); // 인덱스 +1 증가

        if(kind == Kind.STATIC || kind == Kind.FIELD) {
            classRangeTable.put(name, identifier);
        } else {
            subroutineRangeTable.put(name, identifier);
        }
    }

    public static int varCount(Kind kind) {
        if(kind == Kind.STATIC || kind == Kind.FIELD) {
            return (int) classRangeTable.values().stream()
                    .filter(id -> id.getKind() == kind)
                    .count();
        } else {
            return (int) subroutineRangeTable.values().stream()
                    .filter(id -> id.getKind() == kind)
                    .count();
        }
    }

    public static Kind kindOf(String name) {
        Identifier identifier;

        if(CompilationEngine.curRange == Range.SUBROUTINE) {
            identifier = subroutineRangeTable.get(name);
            if(identifier != null) {
                return identifier.getKind();
            }
        }
        identifier = classRangeTable.get(name);
        return (identifier != null) ? identifier.getKind() : Kind.NONE;
    }

    public static String typeOf(String name) {
        Identifier identifier;

        if(CompilationEngine.curRange == Range.SUBROUTINE) {
            identifier = subroutineRangeTable.get(name);
            if(identifier != null) {
                return identifier.getType();
            }
        }
        identifier = classRangeTable.get(name);
        return (identifier != null) ? identifier.getType() : null;
    }

    public static int indexOf(String name) {
        Identifier identifier;

        if(CompilationEngine.curRange == Range.SUBROUTINE) {
            identifier = subroutineRangeTable.get(name);
            if(identifier != null) {
                return identifier.getIndex();
            }
        }
        identifier = classRangeTable.get(name);
        return (identifier != null) ? identifier.getIndex() : -1;
    }
}

File: 11_project/SymbolTable.cpp
#include "SymbolTable.h"

SymbolTable::SymbolTable(void) {
	classTable.clear();
	static_index = 0;
	field_index = 0;

	subroutineTable.clear();
	arg_index = 0;
	var_index = 0;
}

void SymbolTable::startSubroutine(void) {
	map<string, Identifier*>::iterator iter;
	for (iter = subroutineTable.begin(); iter != subroutineTable.end(); iter++)
		delete iter->second;

	subroutineTable.clear();
	arg_index = 0;
	var_index = 0;
}

void SymbolTable::define(string name, string type, iden_kind kind) {
	switch (kind) {
	case STATIC:
		classTable.insert({ name, new Identifier(name, type, kind, static_index) });
		static_index++;
		break;
	case FIELD:
		classTable.insert({ name, new Identifier(name, type, kind, field_index) });
		field_index++;
		break;
	case ARG:
		subroutineTable.insert({ name, new Identifier(name, type, kind, arg_index) });
		arg_index++;
		break;
	case VAR:
		subroutineTable.insert({ name, new Identifier(name, type, kind, var_index) });
		var_index++;
	}
}

int SymbolTable::varCount(iden_kind kind) {
	switch (kind) {
	case STATIC:
		return static_index;
	case FIELD:
		return field_index;
	case ARG:
		return arg_index;
	case VAR:
		return var_index;
	}
}

iden_kind SymbolTable::kindOf(string name) {
	if (in_subroutine) {
		auto item = subroutineTable.find(name);
		if (item != subroutineTable.end())
			return subroutineTable[name]->kind;
	}
	auto item = classTable.find(name);
	if (item != classTable.end())
		return classTable[name]->kind;
	return NONE;
}

string SymbolTable::typeOf(string name) {
	if (in_subroutine) {
		auto item = subroutineTable.find(name);
		if (item != subroutineTable.end())
			return subroutineTable[name]->type;
	}
	auto item = classTable.find(name);
	return classTable[name]->type;
}

int SymbolTable::indexOf(string name) {
	if (in_subroutine) {
		auto item = subroutineTable.find(name);
		if (item != subroutineTable.end())
			return subroutineTable[name]->index;
	}
	auto item = classTable.find(name);
	return classTable[name]->index;
}